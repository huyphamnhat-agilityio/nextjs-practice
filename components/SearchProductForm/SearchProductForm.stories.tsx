import type { Meta, StoryObj } from "@storybook/react";
import { useArgs } from "@storybook/preview-api";

// Components
import SearchProductForm from ".";

const meta: Meta<typeof SearchProductForm> = {
  title: "Components/SearchProductForm",
  component: SearchProductForm,
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof SearchProductForm>;

export const Primary: Story = {};
