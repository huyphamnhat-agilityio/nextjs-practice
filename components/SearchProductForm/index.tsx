"use client";
import { ChangeEvent } from "react";
import { usePathname, useRouter, useSearchParams } from "next/navigation";
import { useDebouncedCallback } from "use-debounce";

// Components
import { SearchIcon } from "@/components/common/Icons";
import { Input } from "../common";

const SearchProductForm = () => {
  const searchParams = useSearchParams();
  const pathname = usePathname();
  const { replace } = useRouter();

  const handleSearch = useDebouncedCallback(
    ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
      const params = new URLSearchParams(searchParams);

      params.set("page", "1");

      if (value) {
        params.set("query", value);
      } else {
        params.delete("query");
      }

      replace(`${pathname}?${params.toString()}`);
    },
    500,
  );
  return (
    <Input
      placeholder="Search courses..."
      startContent={<SearchIcon />}
      classNames={{
        inputWrapper: "max-w-full self-center py-3.75 px-4 ",
      }}
      onChange={handleSearch}
    />
  );
};

export default SearchProductForm;
