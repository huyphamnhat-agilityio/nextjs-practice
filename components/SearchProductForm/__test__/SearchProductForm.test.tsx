import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import SearchProductForm from "@/components/SearchProductForm";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
}));

describe("SearchProductForm", () => {
  const mockReplace = jest.fn();
  const mockUseRouter = useRouter as jest.Mock;
  const mockUsePathname = usePathname as jest.Mock;
  const mockUseSearchParams = useSearchParams as jest.Mock;

  beforeEach(() => {
    mockUseRouter.mockReturnValue({ replace: mockReplace });
    mockUsePathname.mockReturnValue("/");
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  it("should render correctly", () => {
    const { asFragment } = render(<SearchProductForm />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should update the query parameter when the user types", async () => {
    render(<SearchProductForm />);
    const input = screen.getByRole("textbox", {
      name: /search courses\.\.\./i,
    });
    await userEvent.type(input, "React");

    await waitFor(() => {
      expect(mockReplace).toHaveBeenCalledTimes(1);
      expect(mockReplace).toHaveBeenCalledWith("/?page=1&query=React");
    });
  });

  it("should remove the query parameter when the input is cleared", async () => {
    mockUseSearchParams.mockReturnValue(
      new URLSearchParams({ query: "React" }),
    );

    render(<SearchProductForm />);

    const input = screen.getByPlaceholderText("Search courses...");

    await userEvent.clear(input);

    await waitFor(() => {
      expect(mockReplace).toHaveBeenCalledTimes(1);
      expect(mockReplace).toHaveBeenCalledWith("/?page=1");
    });
  });

  it("should debounce the input changes", async () => {
    render(<SearchProductForm />);
    const input = screen.getByRole("textbox", {
      name: /search courses\.\.\./i,
    });

    await userEvent.type(input, "React");

    expect(mockReplace).not.toHaveBeenCalled();

    await waitFor(() => {
      expect(mockReplace).toHaveBeenCalledTimes(1);
      expect(mockReplace).toHaveBeenCalledWith("/?page=1&query=React");
    });
  });
});
