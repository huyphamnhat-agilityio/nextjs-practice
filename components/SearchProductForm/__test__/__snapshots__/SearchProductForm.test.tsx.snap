// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchProductForm should render correctly 1`] = `
<DocumentFragment>
  <div
    class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
    data-filled="true"
    data-filled-within="true"
    data-slot="base"
  >
    <div
      class="h-full flex flex-col"
      data-slot="main-wrapper"
    >
      <div
        class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm gap-3 data-[hover=true]:bg-default-200 group-data-[focus=true]:bg-default-100 min-h-10 rounded-small transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background bg-foreground-200 border-foreground-300 h-auto rounded-1.25 border max-w-full self-center py-3.75 px-4"
        data-slot="input-wrapper"
        style="cursor: text;"
      >
        <div
          class="inline-flex w-full items-center h-full box-border"
          data-slot="inner-wrapper"
        >
          <svg
            fill="currentColor"
            height="30px"
            viewBox="0 0 30 30"
            width="30px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M 13 3 C 7.4889971 3 3 7.4889971 3 13 C 3 18.511003 7.4889971 23 13 23 C 15.396508 23 17.597385 22.148986 19.322266 20.736328 L 25.292969 26.707031 A 1.0001 1.0001 0 1 0 26.707031 25.292969 L 20.736328 19.322266 C 22.148986 17.597385 23 15.396508 23 13 C 23 7.4889971 18.511003 3 13 3 z M 13 5 C 17.430123 5 21 8.5698774 21 13 C 21 17.430123 17.430123 21 13 21 C 8.5698774 21 5 17.430123 5 13 C 5 8.5698774 8.5698774 5 13 5 z"
            />
          </svg>
          <input
            aria-label="Search courses..."
            class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text peer pr-6 rtl:pr-0 rtl:pl-6 group-data-[has-value=true]:text-default-foreground text-foreground-100 h-auto text-sm/7"
            data-has-start-content="true"
            data-slot="input"
            id="react-aria-:r0:"
            placeholder="Search courses..."
            type="text"
            value=""
          />
          <span
            aria-label="clear input"
            class="p-2 -m-2 z-10 hidden absolute right-3 rtl:right-auto rtl:left-3 appearance-none select-none opacity-0 hover:!opacity-100 cursor-pointer active:!opacity-70 rounded-full outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-large peer-data-[filled=true]:opacity-70 peer-data-[filled=true]:block transition-opacity motion-reduce:transition-none"
            data-slot="clear-button"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 24 24"
              width="1em"
            >
              <path
                d="M12 2a10 10 0 1010 10A10.016 10.016 0 0012 2zm3.36 12.3a.754.754 0 010 1.06.748.748 0 01-1.06 0l-2.3-2.3-2.3 2.3a.748.748 0 01-1.06 0 .754.754 0 010-1.06l2.3-2.3-2.3-2.3A.75.75 0 019.7 8.64l2.3 2.3 2.3-2.3a.75.75 0 011.06 1.06l-2.3 2.3z"
                fill="currentColor"
              />
            </svg>
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
